HungarianAlgorithm/Hungarian/Hungarian.cs

```diff
@@ -8,35 +9,206 @@ public class Hungarian
+                while (path is null)
                 {
-                    (0, 3), (2, 3)
-                }),
-                new WellAssignments(1, new List<(int, double)>
+                    if (AreAllNeighboursInSet(eqGraph, S, T, out int? nextT))   // TODO: call this method once and only update bool when S or T is modified
+                    {
+                        // improving potential
+                        var WexceptT = W.Except(T);
+                        double delta = S.Min(s =>
+                            WexceptT.Min(w =>
+                            {
+                                _graph.TryGetEdge(s, w, out var edge);
+                                return edge.Tag - _graph.GetVertexLabel(s) - _graph.GetVertexLabel(w);
+                            })
+                        );
+
+                        // TODO: use wellsSlackness
+                        foreach (int s in S)
+                        {
+                            _graph.AddValueToVertexLabel(s, delta);
+                        }
+                        foreach (int t in T)
+                        {
+                            _graph.AddValueToVertexLabel(t, -delta);
+                        }
+
+                        // Update eqGraph
+                        foreach (int s in S)
+                        {
+                            foreach (var edge in _graph.AdjacentEdges(s))
+                            {
+                                if (!eqGraph.ContainsEdge(edge.Source, edge.Target) &&
+                                    edge.Tag == _graph.GetVertexLabel(edge.Source) + _graph.GetVertexLabel(edge.Target))
+                                {
+                                    eqGraph.AddEdge(new Edge<int>(edge.Source, edge.Target));
+                                    nextT = edge.GetOtherVertex(s);
+                                }
+                            }
+                        }
+                    }
+
+                    // nextT is vertex from (N_p(S) \ T)
+                    var nextTMatchingEdge = Matching.Find(e => e.Source == nextT || e.Target == nextT);
+                    if (nextTMatchingEdge is null)
+                    {
+                        path = new List<Edge<int>>();
+
+                        Edge<int> pathEdge = eqGraph.InEdges(nextT!.Value).First();
+                        while (pathEdge.Source != free_s)
+                        {
+                            path.Add(pathEdge);
+                            pathEdge = eqGraph.InEdges(pathEdge.Source).First();
+                        }
+                        path.Add(pathEdge);
+                    }
+                    else
+                    {
+                        S.Add(nextTMatchingEdge.GetOtherVertex(nextT!.Value));
+                        T.Add(nextT!.Value);
+                    }
+                }
+
+
+                // augment M by path TODO: tidy it up
+                for (int i = 0; i < path.Count / 2; i += 2)
```

Comment:
Ogólnie to nie jest na teraz, ale może lepiej tworzyć zupełnie nowe skojarzenie zamiast usuwać krawędzie?

A jesteś pewien tego iterowania i co 2?

Reply:
Nie pomyślałem w sumie o tym - na pewno do rozważenia, bo to też wiązałoby się z przepisaniem krawędzi, które do skojarzenia już należą, ale nie są ruszane przez `path`.
W sumie jak tak myślę, to może dałoby się trzymać całe skojarzenia w tablicy po prostu, gdzie dla każdego domu trzymamy index przypisanej studni, jeśli jest przypisana...

Iterowanie poprawione 🙂